### ISSUE TITLE
Ability to pipe stdout -> stdin between processes

### ISSUE THREAD

**Issue by ewels** [nextflow-io/nextflow#5028]
This suggestion / request has come up several times, so wanted to collect the [latest thread](https://nextflow.slack.com/archives/C02T98A23U7/p1716565533096219) into a GitHub issue ('cc @nh13 @kmhernan @mahesh-panchal @adamrtalbot @muffato )

Unix pipes are a powerful way to stream data from one process to another, without needing to write intermediate data to disk. Currently this is not possible to do between processes in Nextflow. Current practice is either to write intermediate files, or to put multiple tools into a single process and pipe within that single script block.

Support in Nextflow is not a new request, but is technically challenging for several reasons:

* Making this work with distributed (cloud) clusters
    * May be possible with task batching? https://github.com/nextflow-io/nextflow/pull/3909
* Figuring out how publishing + retry would work

Piping output between containers does work:

Singularity

```bash
singularity exec img1.sif cmd1 | singularity exec img2.sif cmd2
```

Docker

```bash
docker run ubuntu printf "line1\nline2\n" | docker run -i ubuntu grep line2 | docker run -i ubuntu sed 's/line2/line3/g'
```

@mahesh-panchal has written a minimal example demo using named pipes:

<details>

And this was the demo I wrote for using named pipes. The issue there was clean up and potential process deadlock. It would likely work with containers too though:
So just for completion, one can send a pipe, but cleaning up ( i.e. removing the pipe afterwards is not simple because of the working directory isolation).

```nextflow
workflow {
    MKFIFO()
    SENDER( params.message, MKFIFO.out.pipe ) 
    RECEIVER ( MKFIFO.out.pipe ) | view
}

process MKFIFO {
    script:
    """
    mkfifo mypipe
    """

    output:
    path "mypipe", emit: pipe
}

process SENDER {
    input:
    val message
    path pipename

    script:
    """
    echo $message > $pipename
    """

    output:
    path pipename
    
}

process RECEIVER {
    input:
    path pipename

    script:
    """
    cat $pipename
    """

    output:
    stdout
}
```

And this is bad practice as it could easily lead to a process deadlock (edited)
The reason for this structure is because named pipes block until they're read from ( i.e stop the process from completing )

</details>

**Comment by bentsherman** [comment]
Duplicate of #1822

**Comment by bentsherman** [comment]
The first challenge before you even get to the issue of pipes is determining *when* two processes can run on the same node. Even if we allow users to mark when tasks should be grouped, we still have to verify that a configuration is valid.

There is a nice way to make this work, which is an idea I've been toying with for DSL3. If we allow processes to be invoked within an operator closure, instead of baking a `map` operator into into the process, then it becomes very clear when two processes can be grouped. See [this snippet](https://github.com/nf-core/fetchngs/pull/309/files#diff-04efa08d35f4a8f91d2cc70991e6a647787eca3a8ad8116c1242e2e1f0949742) from my fetchngs POC:

```groovy
    sra_metadata
        |> map { meta ->
            sra = SRATOOLS_PREFETCH ( meta )
            SRATOOLS_FASTERQDUMP ( meta, sra )
        }
        |> set { reads }
```

In this example we know that, for each value that comes into the `map` from the input channel, the two tasks can be fused into a single job.

Then it's a matter of whether and how to actually "fuse" the tasks. I think we can take a similar approach to job arrays and task batching by creating a "meta-wrapper" script which calls the `.command.run` of each child task, but in this case we need to chain them rather than simply launching them in parallel, something like:

```bash
TASK1='/work/dir/1'
TASK2='/work/dir/2'

$TASK1/.command.run
# stage outputs of 1 as inputs of 2 ...
$TASK2/.command.run
```

Then the piping can be an enhancement on top of this, where if process 1 has a `stdout` output which matches a `stdin` input to process 2, we can re-jig the wrapper scripts to use a pipe:

```bash
$TASK1/.command.run | $TASK2/.command.run
```

**Comment by bentsherman** [comment]
All that is to say, users should not have to work with named pipes directly. They should just use the `stdout` and `stdin` qualifiers, compose processes in such a way that Nextflow can identify opportunities to pipe and do all of the bash magic under the hood.

**Comment by mahesh-panchal** [comment]
I agree that users shouldn't have to do that. I'm a little bit hesitant on the use of stdout and stdin though given the extra stuff that can end up there. It would be nice if named pipes were just created under the hood, particularly if a file was marked temporary somehow, or maybe streamable.

**Comment by bentsherman** [comment]
That's a good point, I was only thinking of stdout and stdin but really we should be matching the `path` outputs and inputs of the processes.

I don't think we even need to mark the file as streamable. We just need to identify the output-to-input links and manage them with named pipes instead of files. And I think you can duplicate a named pipe in order to e.g. pipe it into a command and also write it to a file, if the user wants to keep it.

**Comment by nh13** [comment]
I’d caution against too much inference being done under the hood. lf I expect that two processes will be run, communicating a pipe, then it should fail if it doesn’t. I.e. I don’t want it to be best effort, I want it to be explicit in the language/operator I use.

**Comment by Poshi** [comment]
Be careful with some ugly software that looks like it could be pipe-friendly, but it isn't. I'm thinking on practices like opening the file to read the magic number, closing the file and then opening the file again to do the actual processing. It tends to happen when you use named pipes where a file is expected (when using stdin it usually works fine).
A way to mark some input/output as streamable (or the oposite, non-streamable) should be present. Otherwise, this kind of software will constantly fail.
Also, be careful when duplicating pipes. Adding a duplication to write a file is fine, but if the duplicated pipe goes to another program, it could cause some performance loss, even a deadlock if the buffers are not big enough and the results of the commands ends in a third command (like in A->B, A->C, B->D, C->D).

**Comment by bentsherman** [comment]
The task chaining by itself is safe and could be automatic, though we'll probably gate it with a config option anyway. The piping is essentially an enhancement atop that, and should be explicitly enabled. Could be a process directive:
```groovy
process foo {
  pipe true
  // ...
}
```

If Nextflow sees that it can chain two tasks, it will only use pipes if both processes have `pipe true`.

The nice thing is that everything we're talking about can be implemented manually in Nextflow. Just combine the processes into one process, and use named pipes as needed. That should make it easy to test different scenarios and see how far we can get with named pipes.

**Comment by muffato** [comment]
I think `pipe` should be a property of each input/output channel. For instance, for an aligner it's likely that its input reference genome would _not_ be `pipe`-able (because random seeks are needed) whereas the input reads would. For outputs, some file formats have a header with information such as the size of the data, etc. Typically, these files are first written with placeholder values and once the data are complete, the writer would seek back to the beginning of the file to complete the header.

**Comment by nh13** [comment]
Seqera containers could enable this. Since you know what combination of packages you need, and you can build the combination on demand for either Docker or Singularity. Then executing them in one container, using traditional Unix pipes, seems easy. If running using conda, then just build an environment with both. And if local, well that’s on you.

**Comment by bentsherman** [comment]
@muffato I see your point. Some files might be amenable to streaming while others aren't. But I wonder if this needs to be defined for both the output and the input, or only the output? In theory, if I have an input stream and I seek to a particular position, the input stream could just block until that much data is received. But I don't know enough about how pipes work in this regard.

@nh13 Yes the software deps are another important point to consider. If both processes specify conda packages then it's easy, just merge them and check for conflicting versions. That can work with or without Wave. But if they use different containers, then you probably have to use Wave, and you probably have to use the conda strategy, although Wave might be able to merge two containers if they have the same base image.

**Comment by bentsherman** [comment]
Fair enough. It just seems to me like Nextflow could provide some mechanism for the input file which can be a stream (if the output is streamed) but fall backs to a file or memory buffer if the downstream task tries to seek through it. That way you wouldn't have to think about whether an input file is streamable, because really it should be able to accept a stream or file. This is similar to how Fusion works.

Such a mechanism is in a weird space between Unix pipes and the Nextflow runtime. Maybe we could do it with some bash magic but I suspect it would be unreliable. Maybe Fusion could do it, even though it's intended mainly for remote storage, but because it is intercepting filesystem calls it could also probably implement a sort of "pipe with fallback".

**Comment by feiloo** [comment]
I think its an antipattern to keep process-scripts in separate nextflow-processes and then recombine them with pipes.
This also requires to ensure piped processes run simultaneously, which is a paradigm break from current nextflow afaict.

**Comment by vinjana** [comment]
> I think its an antipattern to keep process-scripts in separate nextflow-processes and then recombine them with pipes. This also requires to ensure piped processes run simultaneously, which is a paradigm break from current nextflow afaict.

TBH, the piping-feature of Snakemake is the one most important feature for us to choose Snakemake over Nextflow. I therefore completely disagree with your statement, @feiloo, that this would be an antipattern _per se_.

Such a piping achieves the following goals:

1. Such pipes can can be activated and deactivated as performance tuning measure, or to simplify debugging (not implemented in Snakemake, but also not hard to add). In Snakemake the connectivity via pipes is used to automatically infer jobs that have to be co-submitted, which makes it extremely flexible.
2. Instead of having to decide whether jobs need to be connected by piping early during development and then clumsily moving parts between neighboring jobs it is left to later when such performance-tuning is appropriate. Therefore such a pipe feature allows to avoid premature optimization (which is the "root of all evil in programming", according to Donald Knuth). Simply declaring, or not, inputs/outputs as pipes will change the performance architecture of your workflow. Everything in one DSL.
4. There is no convenient DSL in any programming language (AFAIK) for piping processes that is suited for production. For production one needs to track exit-codes, stdout, stderr of each process, otherwise debugging rare problems becomes a nightmare. Mostly workflow developers fall back to crappy solutions, like Bash, which makes the declaration of piped connections easy, but the debugging extremely time-consuming, unless one actually implements the needed features manually, thus blowing up scripts with technical non-business-logic code. Solving this via the DSL would allow for overall clearer code.

For instance, we have some highly tuned legacy workflows that need to be migrated from Bash pipes 😨 into a more a sustainable architecture that makes it easier to run a very large number of workflows (diagnosis, debugging, performance), and also to change the workflows in the future. We also strive to use workflow managers that make it easy to lift research-workflows into production and facilitate a high level of workflow maintainability and long-term improvement. Therefore, the workflow manager should be suited for both, research and production. These are all requirements the usual researcher is not interested in (my experience at least -- only publications count, right?). So we might have special needs in flexibility others don't have.

**Comment by feiloo** [comment]
Pipes already exist in bash.
Duplicating pipes at the nextflow level again for convenience, is an antipattern.
Nextflow is abstractions on top of whole tasks. It runs programs on multiple nodes.
Nextflow pipes would destroy this abstraction barrier.
Nextflows dataflow is coarse grained while pipes would be fine grained. We value resumability, fault tolerance, ... over storage efficiency.

A magical solution for both debugging and performance or research and production can not exist.

A lightweight dsl or tooling to improve/replace bash piping should likely live outside of nextflow.

**Comment by bentsherman** [comment]
The Nextflow language doesn't really impose any assumptions about where and how tasks are executed. Just because I define two processes A and B and then call them in sequence, doesn't mean they have to run on different nodes. The underlying executor should be free to complete these tasks in whatever way it sees fit, including being able to run sequential tasks on the same node to reduce network I/O, or even piping input/output files to reduce disk usage

There is a trade-off between I/O overhead and granularity, but it seems sensible to allow the user to make that trade-off. And if they can do so with a configuration flag rather than a pipeline refactor, all the better, That's exactly the kind of dev experience you would want when your research / production environments have different needs

**Comment by feiloo** [comment]
Our Problem with this feature is that adding support of fine-granular streaming computation will add tons of complexity and bugs and tech debt.

Pipes are special files. Their granularity is tiny (64KB).

Similarily as they color cli programs as "pipeable" and "not-pipeable", they will color nextflow tasks as pipeable and not-pipeable.

Issues include: parallelism (scheduling), random read+writes, resumability, fault-tolerance (retry).

User-managed piping constrains when or where tasks can be executed.
There are traps like: piping over a slow network link, pipe-deadlock,...

Disk-usage problems can be alleviated without pipes (workdir cleanup/garbage collection).
Disk-IO is less a problem since pipeable processes read and write sequentially anyway.
IO optimization is possible without pipes.

Users can already control task granularity (compute) and input/output file granularity through task implementation or parametrization (chunk or shardsizes).

**Comment by Poshi** [comment]
Let's say that we are processing FastQ files that weight 1GB uncompressed.
Let's say too that we have several Nextflow tools:
* GUNZIP, that takes a compressed file and decompresses it
* SPLIT, that splits an input FastQ in several smaller FastQs for further parallelization
* TRIM, that trims some bases from the FastQ
* QUAL_CONVERT, that converts the quality values from PHRED+64 to PHRED+33
* QUAL_TRIM, that trims bases according to their quality
* GZIP, that compresses a file

Let's say that we want to keep these tools independent because otherwise we cannot reuse them in other workflows.
Finally, let's say that we want to run a workflow like:
GUNZIP--->SPLIT-*->TRIM--->QUAL_CONVERT--->QUAL_TRIM--->GZIP
This workflow will take a single FastQ file and write several output files.
Up to this point are the requirements.

If we feed this pipeline a FastQ file, it will write 1GB of data to disk in every step but the last, where it will write the compressed size. Assume the compressed FastQ file size is around 0.5GB. It will write 5*1GB+1*0.5GB==5.5GB.
It will also read a similar amount of data: 1*0.5GB+5*1GB==5.5GB.

With the current Nextflow implementation, it will generate a total I/O load of about 11GB. With an implementation that supports piping, it will only read 0.5G and write 0.5GB for a total I/O load of 1GB. All of that is sequential I/O, but that do not seem relevant to me. It will cause 5.5GB of temporary disk space occupied until the end of the workflow. It will cause 11GB of network traffic to reach the shared file system. It will cause 11GB if raw disk I/O. But it could occupy only 0.5GB of disk space, cause 1GB of network traffic and 1GB of raw disk I/O using pipes.

How can I reduce the I/O load using your approaches, @feiloo? I don't see how workdir cleanup, garbage collection, task granularity or file granularity can help in all of this (chunking the data is something I'm doing in this pipeline, but it does not reduce the I/O load at all).

I agree that this kind of job merging comes with its own issues, specially related to overall resource allocation and deadlocks. But resumability should be the same as in the case of considering all those tools a single one.

**Comment by bentsherman** [comment]
> Similarily as they color cli programs as "pipeable" and "not-pipeable", they will color nextflow tasks as pipeable and not-pipeable.

I don't expect this feature to introduce a process coloring problem. If we add some syntax to mark an input / output as "pipeable", using it it your pipeline code will only denote that the process *can* be piped, not that it must be

There would likely be a configuration option to control whether to enable pipes, so that you can use the same pipeline code for either mode

**Comment by vinjana** [comment]
> Pipes already exist in bash. Duplicating pipes at the nextflow level again for convenience, is an antipattern. Nextflow is abstractions on top of whole tasks. It runs programs on multiple nodes. Nextflow pipes would destroy this abstraction barrier. Nextflows dataflow is coarse grained while pipes would be fine grained. We value resumability, fault tolerance, ... over storage efficiency.

A do agree (again) that an implementation in Nextflow is just one option. I just think that an implementation of such a feature in a workflow management system, that takes already care of proper failure management (no damn `set -e` can be forgotten) and logging (no explicit `1> stdout 2> stderr` redirections polluting the business logic) would provide a clean and easy (for the user) solution.

> A magical solution for both debugging and performance or research and production can not exist.

In Snakemake you can replace the normal `pipe()` annotation with an `optional_pipe()` or so, that you can activate and deactivate dependent on a global debug setting. Nothing magical about that. Just turn on the option, and nothing is piped anymore, all intermediate data is put into files and can be investigated -- which is really handy for debugging the workflow. I think that is sufficient to reject your statement that nothing that "can exist". 
 
> A lightweight dsl or tooling to improve/replace bash piping should likely live outside of nextflow.

TBH, I think it is waste of time to even discuss Bash for any reasonable programming effort. Bash is Turing-complete, yes, but [Brainfuck](https://en.wikipedia.org/wiki/Brainfuck) is either (excuse the rhetoric exaggeration -- of course Bash is much better than Brainfuck, but it is much worse than Python, for example, and IMHO a liability for every production system).

**Comment by Hoeze** [comment]
Any updates on pipes in nextflow? I know of a (number of) projects relying on snakemake due to the lack of piping support in NF
